### ISSUE TITLE
`site` field of `ResolveObjectResponse`

### ISSUE THREAD

**Issue by Sjmarf** [LemmyNet/lemmy#4636]
### Requirements

- [X] Is this a feature request? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a feature request? Do not put multiple feature requests in one issue.
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.
- [X] Do you agree to follow the rules in our [Code of Conduct](https://join-lemmy.org/docs/code_of_conduct.html)?

### Is your proposal related to a problem?

As a client developer, it is difficult to allow users to block instances by domain name. This is because the `BlockInstance` request accepts an `instance_id`, which can only be accessed from a [`Site`](https://join-lemmy.org/api/interfaces/Site.html), which can only be accessed from [`GetCommunityResponse`](https://join-lemmy.org/api/interfaces/GetCommunityResponse.html) or [`GetPersonDetailsResponse`](https://join-lemmy.org/api/interfaces/GetPersonDetailsResponse.html). There is no easy way of getting a `Site` object from a domain name, so it's tricky to find the ID with which to make the `BlockInstance` request.

### Describe the solution you'd like.

A `site` field could be added to [`ResolveObjectResponse`](https://join-lemmy.org/api/interfaces/ResolveObjectResponse.html), allowing a domain name to be converted into a `SiteView`.

### Describe alternatives you've considered.

This process could be used to get the site ID, but it's unideal.

1) Contact the instance that you want to block directly.
2) Use [`GetPosts`](https://join-lemmy.org/api/interfaces/GetPosts.html) with a limit of 1 to get any local post.
3) Take the `actor_id` of the community that the post was made to, and use `ResolveObject` back on the logged-in instance to get the corresponding `CommunityView`.
4) Take the `id` of the returned `CommunityView` and query [`GetCommunity`](`https://join-lemmy.org/api/interfaces/GetCommunity.html`) to get the `Site`.
5) Finally, use the `id` of the `Site` to make a `BlockCommunity` request.

**Comment by dessalines** [comment]
> As a client developer, it is difficult to allow users to block instances by domain name. This is because the [BlockInstance](https://join-lemmy.org/api/interfaces/BlockInstance.html) request accepts an instance_id, which can only be accessed from a [Site](https://join-lemmy.org/api/interfaces/Site.html), which can only be accessed from [GetCommunityResponse](https://join-lemmy.org/api/interfaces/GetCommunityResponse.html) or [GetPersonDetailsResponse](https://join-lemmy.org/api/interfaces/GetPersonDetailsResponse.html). There is no easy way of getting a Site object from a domain name, so it's tricky to find the ID with which to make the BlockInstance request.

Manually typing a domain name in a text field to block, doesn't seem like the best way to block content from a site you don't want to see. Blocking should occur based on things you *see* in the front end. The flow would be:

- See a post or comment on your feed you don't like.
- Have the option to block the entire instance.

This means adding `Site` to more view structs, probably also to the post and comment views.

**Comment by Sjmarf** [comment]
I agree that blocking instances directly from the frontend would be convenient, and that's indeed what we're implementing from community and user pages. If it were to be made easily available for posts and comments API-wise, we would certainly consider adding a "block instance" action there too.

I don't think it's unreasonable to have a "+" button in the list of blocked instances with which a user could add an instance manually. If a user blocks an instance but then also wants to block the instance from another account, a "+" button with which they could enter the domain name manually would be the most convenient option.

**Comment by Nutomic** [comment]
You can get a list of all known domains with their respective instance ids from `v3/federated_instances` and filter that based on user input.

**Comment by Sjmarf** [comment]
Thanks! Didn't think of that :)
